### PR TITLE
Add fused_scale_bias_relu_conv_bnstats OP

### PR DESCRIPTION
**Pull request by Tom-Zheng**
### PR types

New features
### PR changes

OPs
### Description

_Please merge this PR after #54949_

This PR adds `fused_scale_bias_relu_conv_bnstats` which is needed for ResUnit fusion. It is implemented using CUDNN Frontend API.

### CODE REVIEW

**Comment by Tom-Zheng** [comment]
@Xreki For your information, here is the overview of this PR:

**Review changes from last PR #54949:**
- paddle/phi/kernels/autotune/cache_cudnn_frontend.h
- paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h
- paddle/phi/kernels/gpudnn/conv_kernel.cu

**New OP implementation:**

***OP kernel***
The implementation mostly follows CUDNN Frontend example: https://github.com/NVIDIA/cudnn-frontend/blob/12f35fa2be5994c1106367cac2fba21457b064f4/samples/fusion_sample.cpp#L86
- paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_op.cu

***OP registry*** 

- paddle/phi/api/yaml/fused_ops.yaml
- paddle/phi/infermeta/fusion.cc
- paddle/phi/infermeta/fusion.h
- paddle/phi/kernels/CMakeLists.txt

***Unittests*** 

- test/legacy_test/CMakeLists.txt
- test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py

**Comment by Tom-Zheng** [comment]
CI错误后续 @tianshuo78520a 会帮忙手动批准. @Xreki 请麻烦review.

File: paddle/phi/infermeta/fusion.cc [hunk 1]
```diff
@@ -729,4 +729,138 @@ void FastWhereXPUInferMeta(const MetaTensor& condition,
...
+    const std::vector<int>& dilations,
+    const std::vector<int>& strides,
+    const std::string& padding_algorithm,
+    int groups,
+    const std::string& data_format,
+    float momentum,
+    float epsilon,
+    bool fuse_prologue,
+    bool exhaustive_search,
+    int64_t accumulation_count,
```
Comment (Xreki) [inline paddle/phi/infermeta/fusion.cc:843]: 这个参数是什么功能？

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,635 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu]: copyright年份2022 -> 2023

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+DECLARE_bool(cudnn_exhaustive_search);
+
+namespace phi {
+namespace fusion {
+
+using helper = phi::CudnnFrontendConvHelper;
+template <typename T>
+using CudnnDataType = phi::backends::gpu::CudnnDataType<T>;
+
+template <typename T, typename Context>
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:52]: 函数前加一些注释，解释下函数的功能吧

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+          .setNextRunningMeanAndVar(updated_running_mean_desc,
+                                    updated_running_var_desc)
+          .setSavedMeanAndInvVar(saved_mean_desc, saved_inv_var_desc)
+          .setEpsilonTensor(epsilon_desc)
+          .setAccumCountTensor(accum_count_desc)
+          .setExpDecayFactorTensor(exp_decay_desc)
+          .build();
+
+  std::array<cudnn_frontend::Operation const*, 1> ops = {&finalize_stat_op};
+  auto op_graph = cudnn_frontend::OperationGraphBuilder()
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:453]: L449 - L488看着，不同的组合里面，这部分实现是一样的，是否可以封装成一个函数？

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+                        cudnn_version));
+  PADDLE_ENFORCE_GE(dev_ctx.GetComputeCapability(),
+                    80,
+                    phi::errors::PreconditionNotMet(
+                        "This op only supports Ampere and later devices, "
+                        "but got compute capability: %d.",
+                        dev_ctx.GetComputeCapability()));
+  // attr
+  float exp_decay = 1. - momentum;
+  if (epsilon <= CUDNN_BN_MIN_EPSILON - FLT_EPSILON) {
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:518]: 可以用`PADDLE_ENFORCE_LE`

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+  bool deterministic = FLAGS_cudnn_deterministic;
+  PADDLE_ENFORCE_EQ(exhaustive_search && deterministic,
+                    false,
+                    phi::errors::InvalidArgument(
+                        "Cann't set exhaustive_search True and "
+                        "FLAGS_cudnn_deterministic True at same time."));
+  // check optional inputs
+  if (fuse_prologue) {
+    PADDLE_ENFORCE_EQ(
+        scale && bias,
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:536]: scale和bias是`paddle::optional<DenseTensor>&`类型，确定这里这样判断吗？以及后面的打印确定用`%d`吗

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+
+namespace phi {
+namespace fusion {
+
+using helper = phi::CudnnFrontendConvHelper;
+template <typename T>
+using CudnnDataType = phi::backends::gpu::CudnnDataType<T>;
+
+template <typename T, typename Context>
+void _FusedScaleBiasReluConvBnstatsImpl(
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu]: 函数名前的`_`可以去掉

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+}
+
+}  // namespace fusion
+}  // namespace phi
+
+PD_REGISTER_KERNEL(fused_scale_bias_relu_conv_bnstats,
+                   GPU,
+                   ALL_LAYOUT,
+                   phi::fusion::FusedScaleBiasReluConvBnstatsKernel,
+                   phi::dtype::float16) {
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:613]: `bfloat16`类型支持吗？

File: test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
+#   Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Xreki) [inline test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py]: 2022 -> 2023

File: test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            self.y_size[0] * self.y_size[1] * self.y_size[2]
+        )
+
+    def init_attr(self):
+        self.fuse_prologue = True
+        self.exhaustive_search = False
+
+
+@skip_check_grad_ci(reason="no grap op")
+@unittest.skipIf(skip_unit_test(), skip_msg)
```
Comment (Xreki) [inline test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py]: 继承了基类，子类可以不用加`skip`装饰器

File: test/white_list/op_accuracy_white_list.py [hunk 11]
```diff
@@ -92,6 +92,7 @@
 
 NO_FP16_COMPARED_WITH_FP32_OP_LIST = [
     'fake_quantize_moving_average_abs_max',
+    'fused_scale_bias_relu_conv_bnstats',
```
Comment (Xreki) [inline test/white_list/op_accuracy_white_list.py:95]: 为何要把算子加入这个白名单？

File: paddle/phi/infermeta/fusion.cc (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/infermeta/fusion.cc:843]: BatchNorm normalize的元素个数, 单GPU (非SyncBatchNorm)时为`N*H*W`. 详见[Table 42](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnFusedOpsVariantParamLabel_t).

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 8 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:613]: 暂不支持.

File: test/white_list/op_accuracy_white_list.py (hunk 11 above)
Comment (Tom-Zheng) [reply test/white_list/op_accuracy_white_list.py:95]: 如果不加, `check_output_with_place` 会将FP16的结果与FP32结果做对比, 然而该OP不支持FP32输入, 所以会出错.

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 5 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:518]: 这只是个警告, 不会quit. 与batchnorm实现一致: https://github.com/PaddlePaddle/Paddle/blob/42e0c6b800289e43e9c816b819a45552ccdf7215/paddle/phi/kernels/gpu/batch_norm_kernel.cu#L608

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 6 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:536]: 确认过可以work, 它会隐式转为`boolean`

File: test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py (hunk 10 above)
Comment (Tom-Zheng) [reply test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py]: done

File: test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py (hunk 9 above)
Comment (Tom-Zheng) [reply test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py]: done

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 7 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu]: done

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 4 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:453]: done

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 3 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:52]: done

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 2 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu]: done

**Comment by Tom-Zheng** [comment]
@Xreki The changes have commit. Would you please take a look?

**Comment by Tom-Zheng** [comment]
The CI failures need to be manually approved and should not block the review process. @Xreki

**Comment by onecatcn** [comment]
2023-08-08 11:03:40 ****************
2023-08-08 11:03:41 0. You must have Dianhai or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-08-08 11:03:41 1. You must have one RD (XiaoguangHu01,chenwhql,zhiqiu,Xreki,luotao1,qili93,Aurelius84) approval for the usage of const_cast.
2023-08-08 11:03:41 2. Unittest is not allowed to be disabled.
2023-08-08 11:03:41 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-08-08 11:03:41 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-08-08 11:03:41 3. Developers are not allowed to set the check_dygraph field directly, which is set to True by default. If you need to change the check_dygraph field, you must have one RD (phlrain (Recommend), fuyinno4, QingshuChen (Recommend for kunlun) or lanxianghit) review and approve. 
2023-08-08 11:03:41 The code that do not meet the specification are as follows:
2023-08-08 11:03:41  test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py : 
2023-08-08 11:03:41 +                place, atol=self.atol, rtol=self.rtol, check_dygraph=False 
2023-08-08 11:03:41 4. Please use the default precision parameters of 'atol, rtol, eps, max_relative_error'. If you don't use the default value, you must have one RD (Xreki (Recommend), fuyinno4, QingshuChen(Recommend for kunlun), zhiqiu or qili93 (Recommend for NPU) , luotao1, lanxianghit, phlrain or ZzSean or Aurelius84) approval for the usage of other values. The detailed information is in the link: https://github.cor/PaddlePaddle/Paddle/wiki/OP-test-accuracy-requirements. The error line is 
2023-08-08 11:03:41 +                place, atol=self.atol, rtol=self.rtol, check_dygraph=False
2023-08-08 11:03:41 5. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain, qili93, QingshuChen or Aurelius84) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:
2023-08-08 11:03:41 test/legacy_test/test_fused_scale_bias_relu_conv_bnstats_op.py
2023-08-08 11:03:41 +	@skip_check_grad_ci(reason="no grap op")
2023-08-08 11:03:41 There are 6 approved errors.

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,618 @@
...
+  auto compute_dtype = CUDNN_DATA_FLOAT;
+  // create tensor descriptors
+  auto dim_input = phi::vectorize<int64_t>(sum_tensor.dims());
+  std::vector<int64_t> dim_c = {1, dim_input[0], 1, 1};  //  [1, C, 1, 1]
+  std::vector<int64_t> dim_scalar = {1, 1, 1, 1};
+  std::vector<int64_t> stride_scalar = {1, 1, 1, 1};
+
+  std::vector<void*> data_ptrs;
+  std::vector<int64_t> uids;
+  int64_t uid = 100;
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:332]: `uid`是什么，为什么设置成`100`呢？不同`pattern`的`uid`都要设置一样吗？建议`uid`的管理后续可以考虑统一下。

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu [hunk 13]
```diff
@@ -0,0 +1,618 @@
...
+                               &data_ptrs,
+                               &uids,
+                               exhaustive_search,
+                               deterministic,
+                               feature_vector,
+                               &plan_cache);
+}
+
+template <typename T, typename Context>
+void FusedScaleBiasReluConvBnstatsKernel(
```
Comment (Xreki) [inline paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:475]: `FusedScaleBiasReluConvBnstatsKernel`是`FusedScaleBiasReluConvBnstatsImpl`+`BNFinalizeImpl`，所以功能应该是`FusedScaleBiasReluConvBn`？

**Review by Xreki — APPROVED** [review]
LGTM

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 13 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:475]: 是，下个PR修改

File: paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu (hunk 12 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_scale_bias_relu_conv_bnstats_kernel.cu:332]: uid是为了区分不同变量的。不需要一样，只要一个cudnn operation graph内不存在冲突即可。

**Review by wanghuancoder — APPROVED** [review]

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]
LGTM for new fused op

**Review by jzhang533 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@Xreki Would you please merge it?
